### DISCUSSION TITLE
CLI option to clear dangling prune: disabled objects

### DISCUSSION THREAD

**Discussion by cep21** [fluxcd/flux2#2429] [Proposals]
For safety reasons, on a large number of critical namespaces we have to disable prune for flux 2.  This is a fix from a site outage discussed [here](https://github.com/fluxcd/flux2/issues/2421).

This works well, but now eventually we end up with objects that are still around the cluster because they were left by prune.

I would like a CLI command that can detect these objects left via `prune: false` or `kustomize.toolkit.fluxcd.io/prune: disabled` that aren't currently tracked by flux.  It could print them out or just ask me if it's ok to prune them.

**Comment by kingdonb** [comment]
This sounds like a good idea with some potential complications. There are at least two cases of disabled garbage collection to check the solution against, when you do it broadly in an entire Kustomization and when you do it narrowly for an individual resource.

1. `prune: false`
2. annotation `kustomize.toolkit.fluxcd.io/prune: disabled`
3. Also `kustomize.toolkit.fluxcd.io/reconcile: disabled` and potentially others?

(https://fluxcd.io/docs/components/kustomize/kustomization/#garbage-collection) I think the key is to scan every resource in the cluster, or in a particular scope like namespace, and check for any with the label matching your Kustomization(s). Then review the Kustomization inventory to see if the resource is listed or not.

Any unlisted resources that are labeled like this can be subject to potential review and final (remediative) garbage collection:

```
kustomize.toolkit.fluxcd.io/name: parent-kustomization
kustomize.toolkit.fluxcd.io/namespace: namespace-of-ks
```

I don't see any difference between how you handle it by way of which garbage collection disabling method was used, but it seems to me that one of these cases is likely to be easy while the other might be harder. I really don't know enough about the specific implementations of these ideas to say which or why at a glance.

I think this could probably be implemented first on top of the Flux packages as a separate utility, so we can tell whether it will work well or not, then move to incorporate it in the Flux CLI if it is "a Hit."

Thanks for bringing this up! Feel free to implement this as a Proof of Concept and share, I can see this being very useful. 👍
